### DISCUSSION TITLE
CSLA biz obj as an Orleans Grain ??

### DISCUSSION THREAD

**Discussion by swegele** [MarimerLLC/csla#3782] [General]
I don't even know what the heck I'm saying....honestly just thinking about CSLA objects stored/retrieved/created from any app instance from the Orleans cluster.

Yum (drool)

EDIT:  This clicked when [I heard Reuben](https://www.youtube.com/watch?v=tuQFawwwkvI) say that as of Orleans 7.0 there is no more need to inherit from `Grain` base class...you can simply implement `IGrain`.  Got me thinking...

**Comment by rockfordlhotka** [comment]
I don't know a lot about the details of Orleans grains, and specifically how they persist grain state. To persist a CSLA object graph, you'd need to be able to hook the Orleans serialization/deserialization steps and ensure the use of `MobileFormatter`.

**Reply by swegele** [reply]
You've hit it on the head.  When I get a chance I will look further into those extensibility points.  

It seems for objects that get passed **IN** and **OUT** of a grain method, they need the below attributes on the object properties so they can work in their special "high performance" serialization/deserialization process to send things over the network to wherever the grain is physically at (the silo).  But it looks like that is _**NOT**_ required for the "state" objects which are stored inside a grain.

Example:
```cs
[GenerateSerializer]
public class Beer
{
    [Id(0)] public string Id { get; set; } = string.Empty;
    [Id(1)] public string Name { get; set; } = string.Empty;
    [Id(2)] public string Brewery { get; set; } = string.Empty;
    [Id(3)] public string Style { get; set; } = string.Empty;
    ...
}
```

So off the top of my head it looks like you could simply pass in a CSLA object as a `byte[]` and hydrate it inside the grain.  But what about AppContext.  Hmmm.  Like I say, when I get some time I will play with a basic example and see if it is complimentary or just silly pipe dreams.

**Reply by swegele** [reply]
Here is a very well done [youtube vid:](https://www.youtube.com/watch?v=NqM-6pkEqQk)

and [the source code](https://github.com/amolenk/PubCrawlingOrleans) (complex enough to be a real world example) using blazor and signalR etc:
